core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -788,6 +788,20 @@ abstract class RDD[T: ClassTag](
   }
 
   /**
+   * Applies a function to each partition of this RDD, while tracking the index
+   * of the original partition.
+   */
+  def foreachPartitionWithIndex(
```

Comment:
No no, I am not suggesting anything. I am just narrowing the scope the PR by eliminating anything that is not immediately needed in this PR (basic Kafka functionality). There is already a lot more code necessary for this to be viable completely alternative (Driver fault-tolerance, etc.) and it would be good to focus on only that in this PR. The `foreachPartitionWithIndex` seems like something that can be done later independently in a different PR.